File: scripts/integrations/rclone.py
#!/usr/bin/env python3
"""
Cortex rclone integration — universal filesystem connector.

Supports any rclone backend: SFTP, NFS, SMB, Google Drive, OneDrive,
S3-compatible, Backblaze B2, local disk, and more.

The rclone config (containing all remote credentials) is stored encrypted
in the Cortex vault. It is written to a temp file per command and deleted
immediately after — credentials never persist on disk outside the vault.

Setup:
  python scripts/integrations/rclone.py auth

Usage:
  python scripts/integrations/rclone.py auth       [--passphrase <p>]
  python scripts/integrations/rclone.py remotes    [--passphrase <p>]
  python scripts/integrations/rclone.py ls   <remote:path>            [--passphrase <p>]
  python scripts/integrations/rclone.py pull <remote:path> [--dest docs/]  [--passphrase <p>]
  python scripts/integrations/rclone.py push <remote:path> [--src docs/]   [--passphrase <p>]
  python scripts/integrations/rclone.py mount <remote:path> [--mountpoint <path>] [--passphrase <p>]

Requires: rclone installed — https://rclone.org/install/
  macOS:   brew install rclone
  Linux:   sudo apt install rclone  OR  curl https://rclone.org/install.sh | sudo bash
  Windows: winget install Rclone.Rclone
"""

import os
import sys
import argparse
import subprocess
import tempfile

ROOT = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
DOCS_DIR = os.path.join(ROOT, "docs")


# ── rclone check ──────────────────────────────────────────────────────────────

def require_rclone():
    if subprocess.run(["which", "rclone"], capture_output=True).returncode != 0:
        print("ERROR: rclone is not installed.")
        print("  macOS:   brew install rclone")
        print("  Linux:   sudo apt install rclone")
        print("  Windows: winget install Rclone.Rclone")
        print("  Or:      curl https://rclone.org/install.sh | sudo bash")
        sys.exit(1)


# ── Vault helpers ─────────────────────────────────────────────────────────────

def get_secret(name: str, passphrase: str) -> str:
    secrets_script = os.path.join(ROOT, "scripts", "secrets.py")
    result = subprocess.run(
        [sys.executable, secrets_script, "get", name, "--passphrase", passphrase],
        capture_output=True, text=True
    )
    if result.returncode != 0:
        print(f"ERROR: Could not retrieve '{name}' from vault.")
        print("Run: python scripts/integrations/rclone.py auth")
        sys.exit(1)
    return result.stdout.strip()


def store_secret(name: str, value: str, passphrase: str):
    secrets_script = os.path.join(ROOT, "scripts", "secrets.py")
    subprocess.run(
        [sys.executable, secrets_script, "store", name,
         "--value", value, "--passphrase", passphrase],
        check=True
    )


def has_secret(name: str, passphrase: str) -> bool:
    secrets_script = os.path.join(ROOT, "scripts", "secrets.py")
    result = subprocess.run(
        [sys.executable, secrets_script, "get", name, "--passphrase", passphrase],
        capture_output=True, text=True
    )
    return result.returncode == 0


# ── Config temp file ──────────────────────────────────────────────────────────

class RcloneConfig:
    """Context manager: writes rclone config to a temp file, deletes it on exit."""

    def __init__(self, passphrase: str):
        self.passphrase = passphrase
        self._tmp = None

    def __enter__(self) -> str:
        config_data = get_secret("rclone_config", self.passphrase)
        self._tmp = tempfile.NamedTemporaryFile(
            mode="w", suffix=".conf", delete=False, prefix="cortex_rclone_"
        )
        self._tmp.write(config_data)
        self._tmp.flush()
        self._tmp.close()
        return self._tmp.name

    def __exit__(self, *_):
        if self._tmp and os.path.exists(self._tmp.name):
            os.unlink(self._tmp.name)


def run_rclone(args: list, config_path: str) -> int:
    cmd = ["rclone", "--config", config_path] + args
    result = subprocess.run(cmd)
    return result.returncode


# ── Auth ──────────────────────────────────────────────────────────────────────

def cmd_auth(passphrase: str):
    """
    Interactive rclone remote setup. Runs `rclone config` with a temp config,
    then stores the resulting config in the vault.
    """
    require_rclone()

    # Load existing config if present, else start empty
    existing = ""
    if has_secret("rclone_config", passphrase):
        existing = get_secret("rclone_config", passphrase)
        print("Existing rclone config found in vault — adding to it.\n")
    else:
        print("No existing rclone config in vault — starting fresh.\n")

    with tempfile.NamedTemporaryFile(
        mode="w", suffix=".conf", delete=False, prefix="cortex_rclone_"
    ) as tmp:
        tmp.write(existing)
        tmp_path = tmp.name

    try:
        print("Opening rclone config wizard. Add or edit remotes, then quit when done.")
        print("Type 'n' to add a new remote, 'q' to quit.\n")
        subprocess.run(["rclone", "--config", tmp_path, "config"])

        with open(tmp_path, "r") as f:
            new_config = f.read()

        store_secret("rclone_config", new_config, passphrase)
        print("\nrclone config stored in vault.")
        print("Commit cortex.secrets.enc to persist across devices.")
        print("\nRemotes configured:")
        subprocess.run(["rclone", "--config", tmp_path, "listremotes"])
    finally:
        os.unlink(tmp_path)


# ── Remotes ───────────────────────────────────────────────────────────────────

def cmd_remotes(passphrase: str):
    require_rclone()
    with RcloneConfig(passphrase) as cfg:
        print("# Configured remotes\n")
        run_rclone(["listremotes"], cfg)


# ── ls ────────────────────────────────────────────────────────────────────────

def cmd_ls(remote_path: str, passphrase: str):
    require_rclone()
    with RcloneConfig(passphrase) as cfg:
        print(f"# {remote_path}\n")
        run_rclone(["lsf", "--human-readable", remote_path], cfg)


# ── pull ──────────────────────────────────────────────────────────────────────

def cmd_pull(remote_path: str, dest: str, passphrase: str):
    """Copy a file or directory from a remote into a local destination."""
    require_rclone()
    os.makedirs(dest, exist_ok=True)

    with RcloneConfig(passphrase) as cfg:
        print(f"Pulling {remote_path} → {dest}")
        code = run_rclone(["copy", "--progress", remote_path, dest], cfg)

    if code == 0:
        print(f"\nDone. Files are in {dest}")
        print("File this? Add them to docs/ and commit with: git add docs/ && git commit -m 'docs: pull from <remote>'")
    else:
        print(f"\nERROR: rclone exited with code {code}")
        sys.exit(code)


# ── push ──────────────────────────────────────────────────────────────────────

def cmd_push(remote_path: str, src: str, passphrase: str):
    """Push a local directory to a remote (backup)."""
    require_rclone()

    if not os.path.exists(src):
        print(f"ERROR: Source path does not exist: {src}")
        sys.exit(1)

    with RcloneConfig(passphrase) as cfg:
        print(f"Pushing {src} → {remote_path}")
        code = run_rclone(["sync", "--progress", src, remote_path], cfg)

    if code == 0:
        print("\nDone.")
    else:
        print(f"\nERROR: rclone exited with code {code}")
        sys.exit(code)


# ── mount ─────────────────────────────────────────────────────────────────────

def cmd_mount(remote_path: str, mountpoint: str, passphrase: str):
    """Mount a remote as a local filesystem. Desktop/laptop only."""
    require_rclone()
    os.makedirs(mountpoint, exist_ok=True)

    print(f"Mounting {remote_path} at {mountpoint}")
    print("Press Ctrl+C to unmount.\n")

    with RcloneConfig(passphrase) as cfg:
        try:
            run_rclone(["mount", remote_path, mountpoint, "--vfs-cache-mode", "writes"], cfg)
        except KeyboardInterrupt:
            print("\nUnmounted.")


# ── Main ──────────────────────────────────────────────────────────────────────

def prompt_passphrase() -> str:
    import getpass
    return getpass.getpass("Vault passphrase: ")


def main():
    parser = argparse.ArgumentParser(description="Cortex rclone integration — universal filesystem connector")
    parser.add_argument("--passphrase", default=None)
    sub = parser.add_subparsers(dest="cmd")

    sub.add_parser("auth", help="Configure remotes — run once per device")
    sub.add_parser("remotes", help="List configured remotes")

    p_ls = sub.add_parser("ls", help="List files on a remote")
    p_ls.add_argument("remote_path", help="e.g. nas:documents or gdrive:")

    p_pull = sub.add_parser("pull", help="Pull files from a remote into docs/")
    p_pull.add_argument("remote_path", help="e.g. nas:bills/2026")
    p_pull.add_argument("--dest", default=DOCS_DIR, help=f"Local destination (default: {DOCS_DIR})")

    p_push = sub.add_parser("push", help="Push local folder to a remote (backup)")
    p_push.add_argument("remote_path", help="e.g. b2:cortex-backup/docs")
    p_push.add_argument("--src", default=DOCS_DIR, help=f"Local source (default: {DOCS_DIR})")

    p_mount = sub.add_parser("mount", help="Mount a remote as a local filesystem (desktop only)")
    p_mount.add_argument("remote_path", help="e.g. nas:")
    p_mount.add_argument("--mountpoint", default=os.path.expanduser("~/mnt/cortex-remote"),
                         help="Local mountpoint (default: ~/mnt/cortex-remote)")

    args = parser.parse_args()

    if not args.cmd:
        parser.print_help()
        sys.exit(1)

    passphrase = args.passphrase or prompt_passphrase()

    if args.cmd == "auth":
        cmd_auth(passphrase)
    elif args.cmd == "remotes":
        cmd_remotes(passphrase)
    elif args.cmd == "ls":
        cmd_ls(args.remote_path, passphrase)
    elif args.cmd == "pull":
        cmd_pull(args.remote_path, args.dest, passphrase)
    elif args.cmd == "push":
        cmd_push(args.remote_path, args.src, passphrase)
    elif args.cmd == "mount":
        cmd_mount(args.remote_path, args.mountpoint, passphrase)


if __name__ == "__main__":
    main()

File: manifest/framework/scripts/integrations/rclone.ts
#!/usr/bin/env bun
/**
 * Cortex rclone integration — universal filesystem connector.
 *
 * The rclone config is stored encrypted in the Cortex vault. It is written
 * to a temp file per command and deleted immediately after — credentials
 * never persist on disk outside the vault.
 *
 * Setup:
 *   bun manifest/framework/scripts/integrations/rclone.ts auth
 *
 * Usage:
 *   bun manifest/framework/scripts/integrations/rclone.ts [--passphrase <p>] auth
 *   bun manifest/framework/scripts/integrations/rclone.ts [--passphrase <p>] remotes
 *   bun manifest/framework/scripts/integrations/rclone.ts [--passphrase <p>] ls <remote:path>
 *   bun manifest/framework/scripts/integrations/rclone.ts [--passphrase <p>] pull <remote:path> [--dest docs/]
 *   bun manifest/framework/scripts/integrations/rclone.ts [--passphrase <p>] push <remote:path> [--src docs/]
 *   bun manifest/framework/scripts/integrations/rclone.ts [--passphrase <p>] mount <remote:path> [--mountpoint <path>]
 *
 * Requires: rclone installed — https://rclone.org/install/
 */

import { join, dirname } from 'node:path'
import { fileURLToPath } from 'node:url'
import { tmpdir, homedir } from 'node:os'
import { writeFileSync, unlinkSync, mkdirSync } from 'node:fs'
import * as readline from 'node:readline'

const ROOT = join(dirname(fileURLToPath(import.meta.url)), '../..')
const DOCS_DIR = join(ROOT, 'docs')

// --- Arg parser ---

function parseArgs(argv: string[]): Record<string, string | boolean | undefined> {
  const result: Record<string, string | boolean | undefined> = {}
  for (let i = 0; i < argv.length; i++) {
    const arg = argv[i]
    if (arg.startsWith('--')) {
      const key = arg.slice(2)
      const next = argv[i + 1]
      if (next && !next.startsWith('--')) {
        result[key] = next
        i++
      } else {
        result[key] = true
      }
    } else if (!result['_cmd']) {
      result['_cmd'] = arg
    } else if (!result['_pos1']) {
      result['_pos1'] = arg
    }
  }
  return result
}

// --- Interactive prompts ---

async function promptPassphrase(label = 'Vault passphrase'): Promise<string> {
  if (!process.stdin.isTTY) {
    console.error('ERROR: No passphrase provided and stdin is not a TTY. Use --passphrase.')
    process.exit(1)
  }
  return new Promise(resolve => {
    process.stderr.write(`${label}: `)
    let value = ''
    const tty = process.stdin as NodeJS.ReadStream & { setRawMode?: (mode: boolean) => void }
    if (tty.setRawMode) {
      process.stdin.setRawMode(true)
      process.stdin.resume()
      process.stdin.setEncoding('utf8')
      const onData = (ch: string) => {
        if (ch === '\r' || ch === '\n') {
          process.stdin.setRawMode!(false)
          process.stdin.pause()
          process.stdin.removeListener('data', onData)
          process.stderr.write('\n')
          resolve(value)
        } else if (ch === '') {
          process.exit(1)
        } else if (ch === '') {
          value = value.slice(0, -1)
        } else {
          value += ch
        }
      }
      process.stdin.on('data', onData)
    } else {
      const rl = readline.createInterface({ input: process.stdin, output: process.stderr })
      rl.question('', answer => { rl.close(); resolve(answer) })
    }
  })
}

// --- rclone check ---

function requireRclone(): void {
  const proc = Bun.spawnSync(['which', 'rclone'], { stdout: 'pipe', stderr: 'pipe' })
  if (proc.exitCode !== 0) {
    console.error('ERROR: rclone is not installed.')
    console.error('  macOS:   brew install rclone')
    console.error('  Linux:   sudo apt install rclone')
    console.error('  Windows: winget install Rclone.Rclone')
    console.error('  Or:      curl https://rclone.org/install.sh | sudo bash')
    process.exit(1)
  }
}

// --- Vault helpers ---

async function getSecret(name: string, passphrase: string): Promise<string> {
  const proc = Bun.spawn(
    ['bun', join(ROOT, 'manifest/framework/scripts/secrets.ts'), 'get', name, '--passphrase', passphrase],
    { stdout: 'pipe', stderr: 'pipe' }
  )
  const code = await proc.exited
  if (code !== 0) {
    console.error(`ERROR: Could not retrieve '${name}' from vault.`)
    console.error('Run: bun manifest/framework/scripts/integrations/rclone.ts auth')
    process.exit(1)
  }
  return (await new Response(proc.stdout).text()).trim()
}

async function tryGetSecret(name: string, passphrase: string): Promise<string | null> {
  const proc = Bun.spawn(
    ['bun', join(ROOT, 'manifest/framework/scripts/secrets.ts'), 'get', name, '--passphrase', passphrase],
    { stdout: 'pipe', stderr: 'pipe' }
  )
  const code = await proc.exited
  if (code !== 0) return null
  return (await new Response(proc.stdout).text()).trim()
}

async function storeSecret(name: string, value: string, passphrase: string): Promise<void> {
  const proc = Bun.spawn(
    ['bun', join(ROOT, 'manifest/framework/scripts/secrets.ts'), 'store', name, '--value', value, '--passphrase', passphrase],
    { stdout: 'inherit', stderr: 'inherit' }
  )
  const code = await proc.exited
  if (code !== 0) {
    console.error(`ERROR: Failed to store secret '${name}'.`)
    process.exit(1)
  }
}

// --- Config temp file ---

function writeTempConfig(configData: string): string {
  const tmpPath = join(tmpdir(), `cortex_rclone_${Date.now()}.conf`)
  writeFileSync(tmpPath, configData, { mode: 0o600 })
  return tmpPath
}

async function runRclone(args: string[], configPath: string): Promise<number> {
  const proc = Bun.spawn(['rclone', '--config', configPath, ...args], {
    stdout: 'inherit',
    stderr: 'inherit',
    stdin: 'inherit',
  })
  return proc.exited
}

// --- Auth ---

async function cmdAuth(passphrase: string): Promise<void> {
  requireRclone()

  const existing = await tryGetSecret('rclone_config', passphrase)
  if (existing) {
    console.log('Existing rclone config found in vault — adding to it.\n')
  } else {
    console.log('No existing rclone config in vault — starting fresh.\n')
  }

  const tmpPath = writeTempConfig(existing ?? '')
  try {
    console.log("Opening rclone config wizard. Add or edit remotes, then quit when done.")
    console.log("Type 'n' to add a new remote, 'q' to quit.\n")

    const proc = Bun.spawn(['rclone', '--config', tmpPath, 'config'], {
      stdout: 'inherit',
      stderr: 'inherit',
      stdin: 'inherit',
    })
    await proc.exited

    const newConfig = await Bun.file(tmpPath).text()
    await storeSecret('rclone_config', newConfig, passphrase)
    console.log('\nrclone config stored in vault.')
    console.log('Commit cortex.secrets/ to persist across devices.')
    console.log('\nRemotes configured:')
    const listProc = Bun.spawn(['rclone', '--config', tmpPath, 'listremotes'], {
      stdout: 'inherit',
      stderr: 'inherit',
    })
    await listProc.exited
  } finally {
    unlinkSync(tmpPath)
  }
}

// --- Remotes ---

async function cmdRemotes(passphrase: string): Promise<void> {
  requireRclone()
  const configData = await getSecret('rclone_config', passphrase)
  const tmpPath = writeTempConfig(configData)
  try {
    console.log('# Configured remotes\n')
    await runRclone(['listremotes'], tmpPath)
  } finally {
    unlinkSync(tmpPath)
  }
}

// --- ls ---

async function cmdLs(remotePath: string, passphrase: string): Promise<void> {
  requireRclone()
  const configData = await getSecret('rclone_config', passphrase)
  const tmpPath = writeTempConfig(configData)
  try {
    console.log(`# ${remotePath}\n`)
    await runRclone(['lsf', '--human-readable', remotePath], tmpPath)
  } finally {
    unlinkSync(tmpPath)
  }
}

// --- pull ---

async function cmdPull(remotePath: string, dest: string, passphrase: string): Promise<void> {
  requireRclone()
  mkdirSync(dest, { recursive: true })
  const configData = await getSecret('rclone_config', passphrase)
  const tmpPath = writeTempConfig(configData)
  try {
    console.log(`Pulling ${remotePath} → ${dest}`)
    const code = await runRclone(['copy', '--progress', remotePath, dest], tmpPath)
    if (code === 0) {
      console.log(`\nDone. Files are in ${dest}`)
      console.log("File this? Add them to docs/ and commit with: git add docs/ && git commit -m 'docs: pull from <remote>'")
    } else {
      console.error(`\nERROR: rclone exited with code ${code}`)
      process.exit(code)
    }
  } finally {
    unlinkSync(tmpPath)
  }
}

// --- push ---

async function cmdPush(remotePath: string, src: string, passphrase: string): Promise<void> {
  requireRclone()
  const configData = await getSecret('rclone_config', passphrase)
  const tmpPath = writeTempConfig(configData)
  try {
    console.log(`Pushing ${src} → ${remotePath}`)
    const code = await runRclone(['sync', '--progress', src, remotePath], tmpPath)
    if (code === 0) {
      console.log('\nDone.')
    } else {
      console.error(`\nERROR: rclone exited with code ${code}`)
      process.exit(code)
    }
  } finally {
    unlinkSync(tmpPath)
  }
}

// --- mount ---

async function cmdMount(remotePath: string, mountpoint: string, passphrase: string): Promise<void> {
  requireRclone()
  mkdirSync(mountpoint, { recursive: true })
  const configData = await getSecret('rclone_config', passphrase)
  const tmpPath = writeTempConfig(configData)
  console.log(`Mounting ${remotePath} at ${mountpoint}`)
  console.log('Press Ctrl+C to unmount.\n')
  try {
    const code = await runRclone(['mount', remotePath, mountpoint, '--vfs-cache-mode', 'writes'], tmpPath)
    if (code !== 0) {
      console.error(`\nERROR: rclone mount exited with code ${code}`)
    } else {
      console.log('\nUnmounted.')
    }
  } finally {
    try { unlinkSync(tmpPath) } catch { /* already gone */ }
  }
}

// --- Main ---

async function main(): Promise<void> {
  const argv = process.argv.slice(2)
  if (argv.length === 0) {
    console.log('Usage: bun manifest/framework/scripts/integrations/rclone.ts [--passphrase <p>] <subcommand> [options]')
    console.log('Subcommands: auth, remotes, ls, pull, push, mount')
    process.exit(1)
  }

  // Pull out top-level --passphrase before subcommand
  let passphrase: string | undefined
  const rest: string[] = []
  for (let i = 0; i < argv.length; i++) {
    if (argv[i] === '--passphrase' && argv[i + 1]) {
      passphrase = argv[i + 1]
      i++
    } else {
      rest.push(argv[i])
    }
  }

  const cmd = rest[0]
  const cmdArgs = parseArgs(rest.slice(1))

  if (!cmd) {
    console.error('ERROR: No subcommand provided.')
    process.exit(1)
  }

  if (!passphrase) passphrase = await promptPassphrase()

  if (cmd === 'auth') {
    await cmdAuth(passphrase)
  } else if (cmd === 'remotes') {
    await cmdRemotes(passphrase)
  } else if (cmd === 'ls') {
    const remotePath = cmdArgs['_cmd'] as string
    if (!remotePath) { console.error('ERROR: ls requires <remote:path>'); process.exit(1) }
    await cmdLs(remotePath, passphrase)
  } else if (cmd === 'pull') {
    const remotePath = cmdArgs['_cmd'] as string
    if (!remotePath) { console.error('ERROR: pull requires <remote:path>'); process.exit(1) }
    await cmdPull(remotePath, (cmdArgs['dest'] as string) ?? DOCS_DIR, passphrase)
  } else if (cmd === 'push') {
    const remotePath = cmdArgs['_cmd'] as string
    if (!remotePath) { console.error('ERROR: push requires <remote:path>'); process.exit(1) }
    await cmdPush(remotePath, (cmdArgs['src'] as string) ?? DOCS_DIR, passphrase)
  } else if (cmd === 'mount') {
    const remotePath = cmdArgs['_cmd'] as string
    if (!remotePath) { console.error('ERROR: mount requires <remote:path>'); process.exit(1) }
    const defaultMountpoint = join(homedir(), 'mnt', 'cortex-remote')
    await cmdMount(remotePath, (cmdArgs['mountpoint'] as string) ?? defaultMountpoint, passphrase)
  } else {
    console.error(`ERROR: Unknown subcommand '${cmd}'.`)
    process.exit(1)
  }
}

main().catch(err => {
  console.error(`ERROR: ${(err as Error).message}`)
  process.exit(1)
})
